[1 - 3단계 방탈출 사용자 예약] 듀이(이주희) 미션 제출합니다.

ljhee92 opened the pull request:
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  https://prolog.techcourse.co.kr/studylogs/4247

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
- 스프링 학습에 초점을 맞추었기 때문에 객체지향 생활체조 요구사항을 완벽하게 충족하지는 못했습니다. 그렇지만 최대한 의식적으로 지키려고 노력했습니다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [x] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
안녕하세요 로키! BE 7기 듀이입니다.
사용자 예약 미션에서 만나뵙게 되어 반갑습니다!😄
소중한 시간 들여 리뷰해주셔서 감사합니다!
이번 단계 미션의 변경 사항 커밋 목록 링크입니다. [링크](https://github.com/woowacourse/spring-roomescape-member/pull/230/files/93cf7c2e70d1281acbd6adfb6d56fa7769e7792b..588ec1e3bece3abc8621bfc92390c0d70d0f0bf3)

### `Clock` 사용
- 프로덕션 코드에서는 현재 시간을 그대로 사용하고, 테스트 코드에서는 고정된 시간을 주입해주고자 `Clock`을 도입하여 스프링 빈으로 등록해주고 사용했습니다.
- 페어의 권유로 사용해보게 되었는데, 테스트 환경에서의 시간을 독립적으로 활용할 수 있다는 점이 좋았습니다. 또, 스프링 빈으로 등록하는 과정에서 `@Configuration`, `@TestConfiguration`과 같은 어노테이션을 사용할 수 있다는 것도 알게 되었습니다.

### 도메인별 패키지 분리
- 패키지가 `admin`, `global`, `reservation`, `theme`, `time`, `user`로 많이 분리되는 것 같습니다. 도메인별로 패키지를 분리하려고 하다 보니 이렇게 되었는데요.
- `time`과`theme`이 `reservation` 내부로 포함될 수도 있었음에도 불구하고 분리한 이유는 `reservation`과 긴밀한 관계를 가지고 있긴 하지만, `reservation`이 생성되기 전에 `time`과 `theme`이 생성되어야 하기 때문에 독립적이라고 보았기 때문이었습니다.
- 이러한 패키지 분리에 대해 로키는 어떻게 생각하시는지 의견이 궁금합니다.

### `Service`에서 `Service`를 참조할 지, `Repository`를 참조할 지
- `ReservationService`에서 `Reservation` 객체를 만들기 위해(`ReservationService#create`) `ReservationTime`과 `Theme`가 필요하여 각각의 `Service`를 참조했습니다. `Service`를 참조한 이유는 예약시간과 테마를 조회해올 때 존재하는 예약시간, 테마인지 확인하는 로직이 `Service`에 존재했기 때문입니다.
- 그런데 반대로 `ReservationTimeService`와 `ThemeService`에서 예약시간과 테마를 삭제(`ReservationTimeService#delete`, `ThemeService#delete`)할 때엔 예약에 사용된 예약시간과 테마인지 확인하기 위해 `ReservationRepository`를 참조했습니다. `ReservationService`를 참조하게 된다면 순환참조가 일어날 수 있겠다는 페어들의 의견도 있어 이런 구조가 되었는데요.
- `Service`가 다른 여러 `Service`를 참조하기도 하고 여러 `Repository`를 참조하기도 하는지, 올바른 설계인지 잘 모르겠습니다..!

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

solarsalt committed:
docs: add PR template

yeong0jae committed:
feat: 0단계 기본 코드 준비

yeong0jae committed:
docs: 요구사항 명세 정리

yeong0jae committed:
feat: 응답 코드 반환 수정

yeong0jae committed:
feat: 예외처리 추가 및 테스트 리팩터링

yeong0jae committed:
feat: 전역 예외처리 추가

yeong0jae committed:
feat: 테마 crud구현

yeong0jae committed:
test: 테마 crud 테스트 작성

yeong0jae committed:
feat: /reservation 요청이 오면 reservation.html반환

yeong0jae committed:
feat: 예약 가능 시간 조회 API 구현

yeong0jae committed:
feat: 클라이언트 코드 수정

yeong0jae committed:
feat: 테마 랭킹 js 인자 매핑

yeong0jae committed:
feat: 인기 테마 선별 테스트 작성

ljhee92 committed:
docs: 요구사항 명세 수정

ljhee92 committed:
fix: 예약시간 생성 시 상태코드 수정

ljhee92 committed:
refactor: 미사용 메서드 제거

ljhee92 committed:
refactor: 상수 변수명 대문자로 변경, final 키워드 추가, 상수 클래스 생성자 private 제한

ljhee92 committed:
refactor: 라인 컨벤션 통일, 필드 값 final 추가

ljhee92 committed:
refactor: 메서드명 통일

ljhee92 committed:
refactor: DTO 유효성 검증 추가

ljhee92 committed:
refactor: 예외 처리 시 메시지 포함

ljhee92 commented:
@coderabbitai full review

ljhee92 commented:
@coderabbitai full review

ljhee92 commented:
@coderabbitai review `resources`의 프론트엔드 코드 제외하고 리뷰해줘

ljhee92 commented:
@coderabbitai full review resources의 프론트엔드 코드 제외하고 리뷰해줘

Rok93 commented on src/main/java/roomescape/admin/AdminViewController.java:
```diff
@@ -15,11 +15,16 @@ public String getAdmin() {
 
     @GetMapping("/reservation")
```
> 패키지가 admin, global, reservation, theme, time, user로 많이 분리되는 것 같습니다. 도메인별로 패키지를 분리하려고 하다 보니 이렇게 되었는데요.

admin은 별도 도메인이라고 보기에는 조금 어려울 것 같은데요. 🤔
오히려 굳이 따지자면 각각의 API는 각각의 도메인 하위로 분리가 되어야하지 않을까요? 

* `/admin/reservation`: reservation의 controller 패키지
* `/admin/time`: time의 controller 패키지
* `/admin/theme`: theme 의 controller 패키지


> time과theme이 reservation 내부로 포함될 수도 있었음에도 불구하고 분리한 이유는 reservation과 긴밀한 관계를 가지고 있긴 하지만, reservation이 생성되기 전에 time과 theme이 생성되어야 하기 때문에 독립적이라고 보았기 때문이었습니다.
이러한 패키지 분리에 대해 로키는 어떻게 생각하시는지 의견이 궁금합니다.

ljhee92 replied:
로키 말씀대로 `admin`, `user` 패키지는 별도 도메인이 아니다보니 이상하다고 느껴 이런 질문을 하게 됐다고 생각이 됩니다.
로키의 답변을 보고 조금 고민을 해보았는데요!
현재 상황에서는 도메인인 `Reservation`, `ReservationTime`, `Theme`가 각각 독립적인 도메인이 아니고 서로 긴밀히 연관되어 있다고 생각합니다.
그래서 완벽히 도메인별 패키지 분리가 아니었던 기존 구조에서 계층을 먼저 나누는 방식으로 패키지 구조를 변경했습니다.

[refactor: 패키지 구조 변경](https://github.com/ljhee92/spring-roomescape-member/commit/45041c977c287da906b93f04bd2aadf800d47b09)

Rok93 commented on src/main/java/roomescape/time/service/ReservationTimeServiceImpl.java:
```diff
@@ -1,22 +1,35 @@
+
+        LocalTime startAt = request.startAt();
+        if (reservationTimeRepository.existByStartAt(startAt)) {
+            throw new IllegalArgumentException("[ERROR] 이미 존재하는 시간입니다.");
```
예약 시간 혹은 예약 중복 생성에 실패하는 경우 현재 400 응답을 내려주고 있는 것 같은데요. 
더 적절한 [status Code](https://developer.mozilla.org/ko/docs/Web/HTTP/Reference/Status)는 없을까요? 😃

ljhee92 replied:
400 응답의 경우 클라이언트에서 요청이 잘못된 형식으로 들어왔을 때 내려주는 것으로 알고 있는데요!
확실히 예약 시간 혹은 예약 중복 생성에 실패하는 경우는 올바른 형식으로 요청했지만 중복인 경우이기 때문에 400 응답과는 맞지 않다고 생각되네요.
찾아보니 409 응답이 사용자의 요청이 서버의 응답과 충돌하는 경우 사용하는 것으로 400보다 더 적절한 응답이라고 생각되어 변경해주었습니다.

[fix: 예약 시간, 테마, 예약이 중복되어 생성할 수 없는 경우 409 응답 반환하도록 변경](https://github.com/ljhee92/spring-roomescape-member/commit/8b08f2391b707c40fb929537475fc887110291a2)

Rok93 replied:
꼭 409 응답이 아니더라도 다른 Http Status도 한번쯤 훑어봐두면 나중에 적절한 status code를 선택하시는데 도움이되실 것 같아요. 😃

Rok93 replied:
꼭 409 응답이 아니더라도 다른 Http Status도 한번쯤 훑어봐두면 나중에 적절한 status code를 선택하시는데 도움이되실 것 같아요. 😃

Rok93 commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,57 @@
+INSERT INTO theme (name, description, thumbnail)
+VALUES ('레벨2 탈출',
+        '우테코 레벨2를 탈출하는 내용입니다.',
+        'https://example.com/image.jpg');
```
<img width="1913" alt="image" src="https://github.com/user-attachments/assets/d4af7f57-b30a-439b-822b-66b8f8d6e4b1" />

사소하지만 테마 이미지도 조금 더 그럴싸한 것 혹은 정상적인 이미지를 넣어주시면 좋을 것 같아요. 😃

ljhee92 replied:
사소한 부분까지 리뷰해주셔서 감사합니다 로키!
반영해두었습니다😄

[chore: 테마 이미지 추가](https://github.com/ljhee92/spring-roomescape-member/commit/558d450c296c1d2d872c87734b89df147a5c7469)

Rok93 commented on src/main/java/roomescape/reservation/repository/ReservationJdbcRepository.java:
```diff
@@ -38,53 +50,94 @@ public List<Reservation> findAll() {
+
+    @Override
+    public boolean existReservationByTimeId(Long timeId) {
+        String sql = "SELECT COUNT(*) FROM reservation WHERE time_id = ?";
```
단순히 존재 유무 확인이라면 성능 관점에서 `LIMIT 1` 조건을 걸어주거나 혹은 명시적으로 `EXISTS`  키워드를 활용해서 존재여부만 파악하도록 활용해주시는게 좋을 것 같아요. 😃

ljhee92 replied:
앗 넵! 알고 있는 부분이었는데 급하게 구현하다 보니 놓친 것 같습니다.
반영해두었습니다~!

[refactor: 존재 유무 확인 쿼리 수정](https://github.com/ljhee92/spring-roomescape-member/commit/4f24bf687f0e3af7bde23aff184204399ee46a59)

Rok93 commented on src/main/java/roomescape/theme/service/ThemeService.java:
```diff
@@ -0,0 +1,53 @@
+        return ThemeResponse.from(themes);
+    }
+
+    public Theme getTheme(Long id) {
```
사소하지만 단순히 (상태를 꺼내오는 것 같이) get하는 것이 아니라 DB에서 조회하여서 가져오는 것이니 get 보다는 find 같은 네이밍으로 변경해보면 어떨까요? 😃

ljhee92 replied:
오 그렇네요! 의식적으로 생각하지 못했던 부분인데 잡아주셔서 감사합니다
`get`으로 네이밍할 경우 반드시 꼭 하나를 꺼내와야 한다고 생각되고,
`find`로 네이밍한다면 찾아보고 있다면 꺼내온다고 생각되어요.
로키 의견에 동의해서 전체적으로 service, repository의 메서드명 컨벤션과 메서드 순서 통일해두었습니다.

[refactor: 메서드명 컨벤션, 메서드 순서 통일](https://github.com/ljhee92/spring-roomescape-member/commit/daf1cf506848952bcc3f0680ccd513d627b07438)

Rok93 commented on src/main/java/roomescape/theme/service/ThemeService.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    public List<ThemeResponse> getPopularThemes() {
+        List<Theme> themes = themeRepository.findPopularThemeDuringAWeek(10, LocalDate.now(clock));
```
`10`은 매직넘버로 보이는 것 같아요. 😃

ljhee92 replied:
앗 그렇네용!
상수처리 해두었습니다..!

[refactor: 매직넘버 상수처리](https://github.com/ljhee92/spring-roomescape-member/commit/77e0f50ad050644c53f41445b72b06647480ffbb)

Rok93 commented on src/main/java/roomescape/reservation/service/ReservationRepository.java:
```diff
@@ -3,14 +3,24 @@
+import roomescape.reservation.domain.ReserverName;
+import roomescape.theme.domain.Theme;
 
 public interface ReservationRepository {
```
ReservationRepository가 service 패키지에 잘못 위치하는 것 같은데, repository 패키지 하위로 옮겨주세요. 🙏

ljhee92 replied:
페어의 코드로 시작했을 때 페어가 `ReservationRepository`를 의도적으로 사용되는 곳에 두었다고 해서 그렇게 위치시켰었는데요!
패키지 구조를 변경하면서 application 패키지 하위에 두도록 옮겨두었습니다.

[refactor: 패키지 구조 변경](https://github.com/ljhee92/spring-roomescape-member/commit/45041c977c287da906b93f04bd2aadf800d47b09)

Rok93 commented on src/main/java/roomescape/reservation/controller/request/ReservationCreateRequest.java:
```diff
@@ -1,10 +1,14 @@
-        String name,
-        LocalDate date,
-        Long timeId
+        @NotBlank String name,
```
Validation 애노테이션을 활용해주셨군요? 👍👍 

애노테이션 프로퍼티로 message도 지정해주면 어떨까요?😃 
```suggestion
        @NotBlank(message = "예약자명은 필수입니다.") String name,
```

ljhee92 replied:
네 프로퍼티로 message도 지정해주고, ExceptionHandler에서 Validation 예외 잡아서 처리해줄 수 있도록 수정해두었습니다.😊

[refactor: Validation 프로퍼티 message 추가,](https://github.com/ljhee92/spring-roomescape-member/commit/6b9dd904b3534e793d0c2684a8c3112f8639528b) @Valid [누락 추가](https://github.com/ljhee92/spring-roomescape-member/commit/6b9dd904b3534e793d0c2684a8c3112f8639528b)
[refactor: Validation 예외 처리, 에러 응답 JSON으로 수정](https://github.com/ljhee92/spring-roomescape-member/commit/cc511469570b6375eb217dd032afbddb26837ce2)
[chore: ExceptionHandler 패키지 변경, view -> web 변경, Controller 명에서 View, Api 제거](https://github.com/ljhee92/spring-roomescape-member/commit/ac18fc2d5e4deec5d797d8d3e4f5ae86db4e9a77)

Rok93 commented on src/main/java/roomescape/reservation/domain/ReservationDateTime.java:
```diff
@@ -0,0 +1,35 @@
+import roomescape.reservation.domain.exception.PastReservationException;
+import roomescape.time.domain.ReservationTime;
+
+public class ReservationDateTime {
```
ReservationDateTime은 record class로 만들어볼 수 있지 않을까요? 😃

ljhee92 replied:
확실히 `ReservationDateTime`은 `ReservationDate`와 `ReservationTime`을 불변으로 가지고 있고 간단한 유효성 검증만 수행하고 있기 때문에 record class 로 변경해주어도 괜찮을 것 같습니다.
추후에 도메인 로직이 확장될 가능성이 높다면 record class는 적합하지 않을 것 같은데, 그럴 가능성이 현저히 낮다고 생각되어서 보일러 플레이트를 제거할 수 있도록 record class로 변경해주었습니다!

[refactor: ReservationDateTime record class 로 변경](https://github.com/ljhee92/spring-roomescape-member/commit/3c17bc4b73d8226228ccb9528dd3d3c67bfdb33c)

Rok93 commented on src/main/java/roomescape/reservation/domain/ReservationDateTime.java:
```diff
@@ -0,0 +1,35 @@
+    private final ReservationDate reservationDate;
+    private final ReservationTime reservationTime;
+
+    public ReservationDateTime(ReservationDate reservationDate, ReservationTime reservationTime, Clock clock) {
```
순수한 도메인 객체가 프로퍼티로 스프링에 의존적인 `@Bean` 객체를 주입받고있군요 🤔
도메인 객체를 테스트하는데 스프링의 의존성이 생길 것 같다는 생각이 드는데요. 🤔

Clock 객체를 주입받기보다는 생성 시점인 `LocalDateTime`을 주입받는게 조금 더 의존성을 덜어내고 유연하게 객체를 생성할 수 있지 않을까요?

ljhee92 replied:
저 역시도 순수한 도메인 객체인데 스프링 의존성이 생기는 것이 좋지 않다고 생각되네요🤔

도메인인 `ReservationDateTime`에서는 `LocalDateTime`을 주입받도록 하고,
추가 코멘트로 남겨주신 것처럼 `CurrentTimeService`를 두어서 테스트 코드에서는 Fake 객체를 생성하여 테스트할 수 있도록 변경해보았습니다.
`ReservationDateTime` 도메인 테스트 시에도 순수하게 도메인만 테스트할 수 있어서 좋은 것 같습니다👍🏻

[refactor: Clock -> CurrentTimeService로 변경](https://github.com/ljhee92/spring-roomescape-member/commit/b94c137b34e51c5c4ffe7388091b0f585755c671)

Rok93 commented on src/main/java/roomescape/reservation/repository/ReservationJdbcRepository.java:
```diff
@@ -38,53 +50,94 @@ public List<Reservation> findAll() {
-    public Reservation save(Reservation reservation) {
+    @Override
+    public Reservation save(ReserverName reserverName, ReservationDateTime reservationDateTime, Theme theme) {
         SimpleJdbcInsert jdbcInsert = new SimpleJdbcInsert(jdbcTemplate)
```
예약을 새롭게 생성할 때마다 매번 `SimpleJdbcInsert` 객체를 새로 만들고있을 것 같은데요. 🤔
애초에 ReservationJdbcRepository Bean이 생성되는 시점에 SimpleJdbcInsert를 생성해서 상태로 가지게한 뒤, 재사용하도록 변경해보면 어떨까요? 😃

ljhee92 replied:
오! `SimpleJdbcInsert`를 활용하면서 `SimpleJdbcInsert`가 자체적으로 상태를 저장하지는 않기 때문에 재사용이 가능한 것은 모르고 있던 부분이네요😅
매 번 객체 생성할 필요 없이 필드에 두고 Respository 생성 시점에만 생성하고 재사용하도록 변경해두었습니다.
`ReservationJdbcRepository`, `ReservationTimeJdbcRepository`, `ThemeJdbcRepository` 모두 변경해두었어요!

[refactor: SimpleJdbcInsert 객체 재사용하도록 수정](https://github.com/ljhee92/spring-roomescape-member/commit/6ab52ba6bd84b58ef40efb6c9b957f8b3c0ab21e)

Rok93 commented on src/main/java/roomescape/reservation/controller/exception/ReservationExceptionHandler.java:
```diff
@@ -10,12 +10,12 @@
     @ExceptionHandler(ReserverNameEmptyException.class)
-    public ResponseEntity<Void> handleReserverNameEmptyException(ReserverNameEmptyException e) {
-        return ResponseEntity.badRequest().build();
+    public ResponseEntity<String> handleReserverNameEmptyException(ReserverNameEmptyException e) {
```
global 패키지 하위의 전역 예외처리와 reservation 패키지 하위에 전역 예외를 별도로 분리하신 이유가 궁금합니다! 🙋‍♂️

Rok93 replied:
가급적 에러 응답도 단순 문자열을 내려주기보다는 JSON 응답으로 내려주면 좋을 것 같아요. 😃

조금 경험에 의한 조언인데요... 👀
이런 응답들은 가능한 json으로 말아두는 것이 추후 응답을 추가하기에도 좋고 클라이언트단에서는 쌩 문자열을 응답해주게되면 json 파싱이 안돼서 에러가 나는 경우도 종종 있을 수 있어요.

ljhee92 replied:
> global 패키지 하위의 전역 예외처리와 reservation 패키지 하위에 전역 예외를 별도로 분리하신 이유가 궁금합니다! 🙋‍♂️

패키지 구조를 변경하기 전에는 resevation 도메인 내부에서 reservation 예외 처리를 해줘야 하고, global 패키지 하위의 전역 예외처리는 reservation 외에도 reservationTime, theme의 예외처리도 함께 담당하고 있기 때문에 global 패키지 하위에 위치하도록 했습니다.
현재는 패키지 구조를 계층별 분리가 우선되도록 변경한 상태여서, global 패키지에 위치했던 전역 예외처리 클래스를 presentation 패키지 하위의 exception 패키지에 위치하도록 변경해두었습니다.

[chore: ExceptionHandler 패키지 변경, view -> web 변경, Controller 명에서 View, Api 제거](https://github.com/ljhee92/spring-roomescape-member/commit/ac18fc2d5e4deec5d797d8d3e4f5ae86db4e9a77)

> 가급적 에러 응답도 단순 문자열을 내려주기보다는 JSON 응답으로 내려주면 좋을 것 같아요. 😃

경험에 의한 조언 너무 좋습니다👍🏻
저는 아직 경험해보지 못한 부분이라 단순히 String을 반환할 생각밖에 못했는데, 이렇게 짚어주시니 너무 좋습니다!
Validation 예외 처리하면서 다른 예외들도 JSON으로 내려주도록 함께 변경해두었습니다.

[refactor: Validation 예외 처리, 에러 응답 JSON으로 수정](https://github.com/ljhee92/spring-roomescape-member/commit/cc511469570b6375eb217dd032afbddb26837ce2)

Rok93 commented on src/main/java/roomescape/reservation/service/ReservationService.java:
```diff
@@ -28,9 +40,18 @@ public List<ReservationResponse> getAll() {
+        Long timeId = request.timeId();
+        ReservationDate reservationDate = new ReservationDate(request.date());
+
+        if (reservationRepository.existSameDateTime(reservationDate, timeId)) {
+            throw new IllegalArgumentException("[ERROR] 이미 예약이 찼습니다.");
+        }
```
서로 다른 테마일지라도 같은 시간이라면 예약할 수 없는걸까요? 👀
이 부분은 수정한 뒤, 테스트 코드도 추가해주시면 좋을 것 같아요. 😃

ljhee92 replied:
!!! 생각지도 못했던 부분까지 짚어주셔서 감사합니다 로키!
수정하고, 테스트 코드도 추가해두었습니다👍🏻

[fix: 서로 다른 테마는 같은 일시에 예약할 수 있도록 수정](https://github.com/ljhee92/spring-roomescape-member/commit/7d7175a1bfe8a99ea2d9598d85f04a8fceb95e8d)

Rok93 reviewed (changes requested):
듀이 안녕하세요. 
방탈출 사용자 예약 미션을 함께하게된 로키입니다. 잘 부탁드려요. 🤗
1~3단계 미션 잘 구현해주셨네요. 👍👍 
몇몇 코멘트 남겨두었는데, 확인 부탁드릴게요. 🙏 
잘 이해가 안되거나 어려운 부분있으면 언제든지 DM 주세요.  😉

Rok93 commented:
질문에 대한 답변을 빼먹었네요. 😅 
추가 코멘트로 남겨둘게요! 

> Clock 사용
프로덕션 코드에서는 현재 시간을 그대로 사용하고, 테스트 코드에서는 고정된 시간을 주입해주고자 Clock을 도입하여 스프링 빈으로 등록해주고 사용했습니다.
페어의 권유로 사용해보게 되었는데, 테스트 환경에서의 시간을 독립적으로 활용할 수 있다는 점이 좋았습니다. 또, 스프링 빈으로 등록하는 과정에서 @Configuration, @TestConfiguration과 같은 어노테이션을 사용할 수 있다는 것도 알게 되었습니다.

`LocalDateTime.now()`와 같이 제어할 수 없는 요소 때문에 테스트 작성이 어려웠던 코드들을 이런식으로 랩핑하여 쓰는 경우는 이해할 수 있지만, 제가 사용해본 적이 없어서 그런지 Clock을 Bean으로 생성하는 방식은 어떤 장점이 있는건지 잘 모르곘어요. 🤔

저라면 `CurrentTimeService` 같은 서비스 객체를 만들고, 이 객체가 `create()`하면 `LocalDateTime.now()`를 호출하게 했을 것 같아요. 
테스트 코드에서는 이 CurrentTimeService의 페이크 객체를 선언하여 테스트를 작성하기에도 수월하지 않을까 싶

> Service에서 Service를 참조할 지, Repository를 참조할 지

같은 계층끼리 의존하는 것은 말씀하신 것처럼 순환참조 등의 문제가 발생할 여지가 있다고 생각해요. 
Service가 여러 Repository를 의존하는 것은 그렇게 잘못된 것이라 생각되지 않아요. 
개인적으로는 어떤 하나의 유즈케이스를 수행하기위해서는 여러 (DB에 저장된) 객체들이 필요할 수 있기 때문에 자연스러운 현상이라고 생각해요. 

결국 같은 계층끼리 서로 의존하는 상황만 지양한다면 저는 문제될 것 없다고 생각해요.

ljhee92 committed:
refactor: 패키지 구조 변경

- Reservation, ReservationTime, Theme 도메인들이 완벽히 독립적이지 않고 연관되어 있기 때문에 도메인 기반 내부 계층 설계보다 계층형 설계로 변경

ljhee92 committed:
refactor: 존재 유무 확인 쿼리 수정

COUNT(*)로 확인할 경우 존재하는 모든 행을 확인하고 개수를 계산하므로 검색할 개수를 1개로 제한하여 찾도록 수정

ljhee92 committed:
chore: 테마 이미지 추가

ljhee92 committed:
refactor: 메서드명 컨벤션, 메서드 순서 통일

ljhee92 committed:
refactor: 매직넘버 상수처리

ljhee92 committed:
refactor: Validation 프로퍼티 message 추가, @Valid 누락 추가

ljhee92 committed:
refactor: Validation 예외 처리, 에러 응답 JSON으로 수정

ljhee92 committed:
chore: ExceptionHandler 패키지 변경, view -> web 변경, Controller 명에서 View, Api 제거

ljhee92 committed:
refactor: ReservationDateTime record class 로 변경

- ReservationDate와 ReservationTime을 불변으로 가지고 있고, 간단한 유효성 검증만 진행하고 있으며 추후 확장 가능성이 낮다고 판단하여 보일러 플레이트 감소를 위해 record class로 변경

ljhee92 committed:
refactor: SimpleJdbcInsert 객체 재사용하도록 수정

- SimpleJdbcInsert는 내부적으로 SQL을 생성하지만 자체적으로 상태를 저장하고 있지는 않기 때문에 재사용이 가능하므로 매 번 객체 생성을 할 필요 없이 필드에 두도록 수정

ljhee92 committed:
fix: 서로 다른 테마는 같은 일시에 예약할 수 있도록 수정

ljhee92 committed:
fix: 예약 시간, 테마, 예약이 중복되어 생성할 수 없는 경우 409 응답 반환하도록 변경

ljhee92 committed:
refactor: Clock -> CurrentTimeService로 변경

ljhee92 committed:
refactor: 구현체가 아닌 interface 의존하도록 수정

ljhee92 committed:
refactor: Override 어노테이션 추가

ljhee92 committed:
fix: RowMapper 매핑 시 컬럼명 일치, 중복 Mapper 상수 처리

ljhee92 commented:
안녕하세요 로키!
피드백 주신 부분 꼼꼼하게 반영하려고 노력해보았습니다!
코드래빗이 해준 리뷰도 반영해볼 수 있는 것들은 함께 반영해보고 resolve 처리해두기도 했는데 여전히 많네요..😅
다 반영하다가는 너무 늦어질 것 같아서 로키의 피드백을 최우선으로 반영하고 요청드립니다.
연휴인데 리뷰 요청드리게 되어 죄송하네용🥹
항상 시간 들여 리뷰해주셔서 감사하게 생각합니다!👍🏻

Rok93 commented on src/main/java/roomescape/presentation/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,40 @@
+
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public ResponseEntity<ErrorResponse> handleMethodArgumentNotValidException(MethodArgumentNotValidException e) {
+        String message = e.getBindingResult()
```
Spring Validation 애너테이션에 message 프로퍼티 설정만 말씀드렸는데, 예외 응답 메시지도 잘 출력해주셨군요? 👍💯

Rok93 commented on src/main/java/roomescape/presentation/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,40 @@
+import roomescape.presentation.dto.response.ErrorResponse;
+
+@RestControllerAdvice
+public class GlobalExceptionHandler {
```
예외처리를 하지않은 부분에서 발생한 예외들은  서버 에러(5xx)로 볼 수도 있을 것 같은데요. 
전역 예외처리에서 다른 예외처리에서 캐치하지 못한 예외들은 캐치해서 500 에러응답을 내려주도록 해보면 어떨까요? 😃

ljhee92 replied:
오 그렇네요!
처리하지 못한 예외가 생겼을 때 `GlobalExceptionHandler`에서 잡아서 500 에러 응답 내려줄 수 있도록 해두었습니다👍🏻

[feat: 캐치하지 못한 예외들 처리 추가](https://github.com/ljhee92/spring-roomescape-member/commit/281a7ade3a86fbc9d76894e2ac1032012821819c)

Rok93 commented on src/test/java/roomescape/domain/ReservationDateTimeTest.java:
```diff
@@ -0,0 +1,36 @@
+
+public class ReservationDateTimeTest {
+
+    @Test
+    void 현재_이전_시간에는_예약할_수_없다() {
+        LocalDateTime now = LocalDateTime.now();
+        LocalDate today = now.toLocalDate();
+        ReservationDate reservationDate = new ReservationDate(today);
+        LocalTime time = now.toLocalTime().minusMinutes(1);
+        ReservationTime reservationTime = ReservationTime.create(time);
+
+        assertThatThrownBy(() -> ReservationDateTime.create(reservationDate, reservationTime, now))
+                .isInstanceOf(PastReservationException.class);
+    }
```
토끼의 의견과 동일한데요. 
가능한 테스트 코드에는 고정된 날짜 값을 사용하는 것을 추천해요. 
테스트할 때마다 인자 값이 바뀌면 일관된 결과를 보장하기도 어려울 수 있다고 생각해요. 🤔

ljhee92 replied:
Fake 객체로 시간을 고정해두었으면서 도메인에서는 활용하지 못했었네요!
앞으로 테스트 시에는 시간을 고정해서 테스트할 수 있도록 좀 더 주의해야겠다고 생각했습니다😅
반영해두었습니다!

[refactor: 테스트 시간 고정](https://github.com/ljhee92/spring-roomescape-member/commit/3c4d7fafb166077ccca54ba1509886423f4ffd81)

Rok93 commented on src/test/java/roomescape/application/ReservationServiceTest.java:
```diff
@@ -0,0 +1,143 @@
+                theme.getId()
+        ));
+        List<ReservationResponse> responses = reservationService.getReservations();
+        ReservationResponse response = responses.get(0);
+
+        assertThat(response.id()).isEqualTo(1L);
+        assertThat(response.name()).isEqualTo(ReserverNameFixture.한스.getName());
+        assertThat(response.date()).isEqualTo(ReservationDateFixture.예약날짜_25_4_22.getDate());
+        assertThat(response.time()).isEqualTo(
+                new ReservationTimeResponse(reservationTime.getId(), reservationTime.getStartAt().toString()));
```
다건 조회 케이스는 조회 결과의 size도 검증해주는 것이 좋다고 생각해요. 😃
```suggestion

        assertThat(response.size).isEqualTo(1);
        assertThat(response.get(0).id()).isEqualTo(1L);
        assertThat(response.get(0).name()).isEqualTo(ReserverNameFixture.한스.getName());
        assertThat(response.get(0).date()).isEqualTo(ReservationDateFixture.예약날짜_25_4_22.getDate());
        assertThat(response.get(0).time()).isEqualTo(
                new ReservationTimeResponse(reservationTime.getId(), reservationTime.getStartAt().toString()));
```

ljhee92 replied:
꼼꼼하게 리뷰해주셔서 감사해요 로키!
말씀해주신 내용대로 다건 조회 케이스는 조회 결과의 size까지 검증해주는 것이 명확한 테스트가 되겠네요!
`ReservationTimeService`, `ThemeService`의 조회 테스트에서도 추가해주었습니다.😄

[refactor: 다건 조회 케이스 size 검증 추가](https://github.com/ljhee92/spring-roomescape-member/commit/0de9ef6a0d00864bc2ce5a2119fe70d774cd0f17)
[refactor: 여러 개의 테스트 assertAll로 수정](https://github.com/ljhee92/spring-roomescape-member/commit/b74736463cc9bf6728dc944b3bc3cebfd7ae1578)

Rok93 commented on src/test/java/roomescape/application/ReservationServiceTest.java:
```diff
@@ -0,0 +1,143 @@
+        ReservationTime reservationTime = reservationTimeDbFixture.예약시간_10시();
+        Theme theme = themeDbFixture.공포();
+
+        reservationService.createReservation(new ReservationCreateRequest(
+                ReserverNameFixture.한스.getName(),
+                ReservationDateFixture.예약날짜_25_4_22.getDate(),
+                reservationTime.getId(),
+                theme.getId()
+        ));
+        reservationService.deleteReservationById(1L);
```
어떤 id 값이 나올지 추측해서 코드를 테스트 코드를 짜는 것 보다는,  
생성한 결과에서 id를 추출해서 삭제 요청에 넣으면 되지 않을까요? 

추가적으로 test 코드또한 유지보수의 대상이므로 가능하면  테스트 코드 작성 시 given, when, then  구조를 한눈에 파악할 수 있도록 작성하는 연습을 해보시면 좋을 것 같아요. 

예를들어 지금의 테스트 케이스는 어떤 메서드를 테스트하려는 것인지 한눈에 들어오지가 않는 것 같아요. 👀

```suggestion
        var createdReservation = reservationService.createReservation(new ReservationCreateRequest(
                ReserverNameFixture.한스.getName(),
                ReservationDateFixture.예약날짜_25_4_22.getDate(),
                reservationTime.getId(),
                theme.getId()
        ));
        reservationService.deleteReservationById(createdReservation.id());
```

ljhee92 replied:
맞아요! 로키의 의견에 동의합니다.
Fixture를 여러 개 도입해서 사용하는 것은 처음 시도해본 것이어서 정신없이 테스트 코드를 짰던 것 같아요.🥲
생성한 객체의 id를 추출해서 삭제 요청에 넣도록 수정해두었습니다.
given, when, then 구조를 조금 더 명확히하는 리팩터링은 아래 피드백에 함께 반영했습니다!

[refactor: 생성된 결과에서 id를 추출하여 삭제하도록 수정](https://github.com/ljhee92/spring-roomescape-member/commit/4c5320dac4482f9bdd1b73e43d799d2dc985577c)
https://github.com/woowacourse/spring-roomescape-member/pull/230#discussion_r2076788189

Rok93 commented on src/test/java/roomescape/application/ReservationServiceTest.java:
```diff
@@ -0,0 +1,143 @@
+        ReservationTime reservationTime = reservationTimeDbFixture.예약시간_10시();
+        Theme theme = themeDbFixture.공포();
+
+        reservationService.createReservation(new ReservationCreateRequest(
+                ReserverNameFixture.한스.getName(),
+                ReservationDateFixture.예약날짜_25_4_22.getDate(),
+                reservationTime.getId(),
+                theme.getId()
+        ));
```
**ReservationServiceTest** 클래스는  `reservationService` 객체에 대한 테스트를 하기위한 테스트 클래스인데요. 
fixture 생성을 이 객체의 메서드로하는 것은 조금 찝찝하지 않나요? 🤔

개인적으로 검증하려는 대상을 테스트의 given 절에서 사용한다는게 조금 어색하다고 생각되는데요. 👀
given 절은 조금 더 신뢰할 수 있는 방식으로 생성해보면 어떨까요? (ex. JdbcTempalte, SimpleJdbcInsert)

ljhee92 replied:
`ReservationService`를 테스트하고 있으면서 `ReservationService#createReservation`을 사용하여 객체를 생성하는 것이 찜찜합니다😭
`Reservation` 자체도 Fixture를 활용할 수 있도록 수정해보았습니다.

[refactor: given절의 Reservation을 Fixture로 수정](https://github.com/ljhee92/spring-roomescape-member/commit/ea0dfde59f10043236d0cd520572b73d9a0ba85c)

Rok93 commented on src/test/java/roomescape/fixture/ReservationDbFixture.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    public Reservation 예약_한스_25_4_22_10시_공포(ReservationTime reservationTime, Theme theme) {
+        SimpleJdbcInsert jdbcInsert = new SimpleJdbcInsert(jdbcTemplate)
```
이 부분도 동일하게 `SimpleJdbcInsert` 객체를 매번 만들 필요 없이, 객체 생성 시점에 필드로 선언해두고 재사용하면 되지 않을까요? 😃

ljhee92 replied:
피드백 하나하나 로키의 꼼꼼함에 감탄하고 있습니다..👍🏻
테스트 Fixture들에 있는 `SimpleJdbcInsert` 객체들도 재사용할 수 있도록 수정했습니다.

[refactor: SimpleJdbcInsert 객체 재사용하도록 수정](https://github.com/ljhee92/spring-roomescape-member/commit/ac10e3310044d0a8a077d14acdf1462b12604e77)

Rok93 reviewed (approved):
듀이 피드백 반영 잘해주셨어요. 👍👍 
1~3단계는 충분히 잘 수행해주신 것 같아서, 이만 머지하도록하겠습니다. 😉
추가로 코멘트 남겨두었는데, 다음 미션 진행하시면서 같이 확인해서 반영 부탁드릴게요. 🤗

ljhee92 commented on src/main/java/roomescape/presentation/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,40 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+import roomescape.presentation.dto.response.ErrorResponse;
+
+@RestControllerAdvice
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler(NoSuchElementException.class)
+    public ResponseEntity<Void> handleNoSuchElementException(NoSuchElementException e) {
+        return ResponseEntity.notFound().build();
+    }
+
+    @ExceptionHandler(IllegalArgumentException.class)
+    public ResponseEntity<ErrorResponse> handleIllegalArgumentException(IllegalArgumentException e) {
+        String message = e.getMessage();
+        ErrorResponse errorResponse = ErrorResponse.of(HttpStatus.CONFLICT, message);
+        return ResponseEntity
+                .status(HttpStatus.CONFLICT)
+                .body(errorResponse);
+    }
+
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public ResponseEntity<ErrorResponse> handleMethodArgumentNotValidException(MethodArgumentNotValidException e) {
+        String message = e.getBindingResult()
+                .getFieldErrors().stream()
+                .map(DefaultMessageSourceResolvable::getDefaultMessage)
+                .collect(Collectors.joining(", "));
+        ErrorResponse errorResponse = ErrorResponse.of(HttpStatus.BAD_REQUEST, message);
+        return ResponseEntity.badRequest().body(errorResponse);
+    }
+}
```
캐치하지 못한 예외들 이렇게 처리했습니다~!

[feat: 캐치하지 못한 예외들 처리 추가](https://github.com/ljhee92/spring-roomescape-member/commit/281a7ade3a86fbc9d76894e2ac1032012821819c)
